snuba/subscriptions/scheduler.py

```diff
@@ -0,0 +1,43 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Generic, Iterator, TypeVar
+
+from snuba.subscriptions.types import TTimestamp
+from snuba.utils.types import Interval
+
+
+TTask = TypeVar("TTask")
+
+
+@dataclass(frozen=True)
+class ScheduledTask(Generic[TTimestamp, TTask]):
+    """
+    A scheduled task represents a unit of work (a task) that is intended to
+    be executed at (or around) a specific time.
```

Comment:
Does this mean that for recurrent tasks (like every minute) we will have an instance of ScheduledTask per minute ?
This does not represent a line a crontab file, right ?

Reply:
Yeah, for example, if the scheduler contains some state that more or less says "execute task T every minute", if we call `find` on the scheduler with an interval representing 00:00:00 to 00:05:00, it should yield something like:

```
[
  [00:01:00, T],
  [00:02:00, T],
  [00:03:00, T],
  [00:04:00, T],
  [00:05:00, T],
]
```

So the scheduler contains/maintains the equivalent of the crontab, but provides access to the materialized schedule entries themselves, if that makes sense.

Would it help to elaborate this in the docstring?

Reply:
_Would it help to elaborate this in the docstring?_

Won't hurt, but that also depends if you want to make this an implementation decision.

Reply:
Going to address improving this later as designs stabilize. Please hold me to that.

Reply:
Going to address improving this later as designs stabilize. Please hold me to that.